src/Lecture4.hs

```diff
@@ -226,18 +327,49 @@ implement the next task.
 -}
 
 combineRows :: NonEmpty Row -> Stats
-combineRows = error "TODO"
+combineRows rows = case rows of
+    x :| [] -> rowToStats x
+    (x :| xs) -> foldl' (<>) (rowToStats x) (Prelude.map rowToStats xs)
 
 {-
 After we've calculated stats for all rows, we can then pretty-print
 our final result.
 
+    let stringStats =
+            [ "Total positions        : 3"
+            , "Total final balance    : -15"
+            , "Biggest absolute cost  : 50"
+            , "Smallest absolute cost : 10"
+            , "Max earning            : 25"
+            , "Min earning            : 10"
+            , "Max spending           : 50"
+            , "Min spending           : 50"
+            , "Longest product name   : Pineapples"
+            ]
 If there's no value for a field (for example, there were not "Buy" products),
 you can return string "no value".
 -}
 
+
+
 displayStats :: Stats -> String
-displayStats = error "TODO"
+displayStats stats = "Total positions        : " <> show (getSum (statsTotalPositions stats))
+                    <> "\nTotal final balance    : " <> show (getSum (statsTotalSum stats))
+                    <> "\nBiggest absolute cost  : " <> show (getMax (statsAbsoluteMax stats))
+                    <> "\nSmallest absolute cost : " <> show (getMin (statsAbsoluteMin stats))
+                    <> "\nMax earning            : " <> case statsBuyMax stats of
+                                                        Nothing -> "no value"
+                                                        _ -> show (getMax (fromJust (statsSellMax stats)))
```

Comment:
You can use a standard function `maybe`:

* https://hackage.haskell.org/package/base-4.17.1.0/docs/Data-Maybe.html#v:maybe

The code will look like this:

```suggestion
                    <> "\nMax earning            : " <> maybe "no value" (show . getMax) $ statsBuyMax stats
```

You can then later move the common parts into a local helper function if you feel so 🙂